src/master-playlist-controller.js

```diff
@@ -1325,38 +1339,44 @@ export class MasterPlaylistController extends videojs.EventTarget {
    * indefinitely.
    * @private
    */
-  excludeIncompatibleVariants_(media) {
-    let codecCount = 2;
-    let videoCodec = null;
-    let codecs;
-
-    if (media.attributes.CODECS) {
-      codecs = parseCodecs(media.attributes.CODECS);
-      videoCodec = codecs.videoCodec;
-      codecCount = codecs.codecCount;
-    }
+  excludeIncompatibleVariants_(codecString) {
+    const codecs = parseCodecs(codecString);
+    const codecCount = Object.keys(codecs).length;
 
-    this.master().playlists.forEach(function(variant) {
-      let variantCodecs = {
-        codecCount: 2,
-        videoCodec: null
-      };
+    this.master().playlists.forEach((variant) => {
+      // skip variants that are already blacklisted forever
+      if (variant.excludeUntil === Infinity) {
+        return;
+      }
+      // TODO: should we really assume codecs?
+      let variantCodecs = {};
+      let variantCodecCount = 2;
 
       if (variant.attributes.CODECS) {
         variantCodecs = parseCodecs(variant.attributes.CODECS);
+        variantCodecCount = Object.keys(variantCodecs).length;
+      }
+
+      // The number of streams cannot change
+      if (variantCodecCount !== codecCount) {
+        variant.excludeUntil = Infinity;
       }
 
-      // if the streams differ in the presence or absence of audio or
-      // video, they are incompatible
-      if (variantCodecs.codecCount !== codecCount) {
+      // the video codec cannot change
+      if (variantCodecs.video && codecs.video &&
+        variantCodecs.video.type.toLowerCase() !== codecs.video.type.toLowerCase()) {
         variant.excludeUntil = Infinity;
       }
 
-      // if h.264 is specified on the current playlist, some flavor of
-      // it must be specified on all compatible variants
-      if (variantCodecs.videoCodec !== videoCodec) {
+      // the audio codec cannot change
+      if (variantCodecs.audio && codecs.audio &&
+        variantCodecs.audio.type.toLowerCase() !== codecs.audio.type.toLowerCase()) {
         variant.excludeUntil = Infinity;
       }
+
+      if (variant.excludeUntil === Infinity) {
+        this.logger_(`blacklisting ${variant.id} as variant codecs ${variant.attributes.CODECS} are incompatible with ${codecString}`);
```

Comment:
This might read a bit funny if there is no CODECS attribute. Maybe we should have a special case log for that?